### PR TITLE
Functional Test Drivers Overhaul

### PR DESCRIPTION
**Pull request by jmchilton**
- Re-introduce the ability to target certain tests against an external Galaxy instance - with `GALAXY_TEST_EXTERNAL`.
- Started some documentation for all the various environment variables used by the testing infrastructure.
- Large de-duplication effort aimed at reducing overlap between tool shed and Galaxy test drivers.
- Added documentation to all existing and new public methods in these files.
- De-duplication of environment variables means one can now prepopulate tool shed tests like with Galaxy using `TOOL_SHED_TEST_DB_TEMPLATE` that mirrors the behavior of `GALAXY_TEST_DB_TEMPLATE`.
- Eliminated a bunch of unused code and some likely unused code.
- Fixed some method names to use Python naming conventions.
- This work has greatly reduced the cyclomatic complexity of the test driver code - for instance the Mccabe complexity of the `main` function in `scripts/functional_tests.py` has been reduced from 64 to 3.

### CODE REVIEW

**Comment by guerler** [comment]
@galaxybot test this

File: run_tests.sh [hunk 1]
```diff
@@ -35,6 +35,14 @@ Run a specific API test:
     ./run_tests.sh -api test/api/test_tools.py:ToolsTestCase.test_map_over_with_output_format_actions
 
 
+External Tests:
+
+A small subset of tests can be run against an existing Galxy
+instance. The external Galaxy instance URL can be configured with
+--external_url. If this is set eithre --external_master_key or
```
Comment (nsoranzo) [inline run_tests.sh]: s/set eithre/set, either/

File: run_tests.sh [hunk 2]
```diff
@@ -45,6 +53,88 @@ Extra options:
...
                        or 'sqlite' databases.
+  --external_url       External URL to use for Galaxy testing (only certain tests).
+  --external_master_key Master API key used to configure external tests.
+  --external_user_key  User API used for external tests - not required if
+                       external_master_key is specified.
+
+Environment Variables:
+
+In addition to the above command-line options, many environment Variables
+can be used to control the Galaxy functional testing processing.
```
Comment (nsoranzo) [inline run_tests.sh]: We should probably state that command-line options take precedence over environment variables (even if this is common practice).

File: run_tests.sh [hunk 3]
```diff
@@ -45,6 +53,88 @@ Extra options:
...
  --db <type>           For use with --dockerize, run tests using partially migrated 'postgres', 'mysql',
                        or 'sqlite' databases.
+  --external_url       External URL to use for Galaxy testing (only certain tests).
+  --external_master_key Master API key used to configure external tests.
+  --external_user_key  User API used for external tests - not required if
+                       external_master_key is specified.
+
+Environment Variables:
+
+In addition to the above command-line options, many environment Variables
```
Comment (nsoranzo) [inline run_tests.sh]: s/Variables/variables/

**Comment by jmchilton** [comment]
@nsoranzo Thanks for the eagle eyes - I have rebased the last little doc commit (now b8ecf36) with your suggested changes.

**Comment by jmchilton** [comment]
Thanks for the merge @guerler and thanks again for the comments @nsoranzo.

**Pull request merged**